File: src/main/java/io/myosotisdev/utopianism/battle/DamageModifier.java
package io.myosotisdev.utopianism.battle;

import io.myosotisdev.utopianism.util.IntegerTemplate;

public class DamageModifier
{
    private String source;
    private DamageType type;
    private IntegerTemplate value;
    private boolean ignoreShield;

    public DamageModifier(DamageType type, int value)
    {
        this(type, value, false);
    }

    public DamageModifier(DamageType type, int value, boolean ignoreShield)
    {
        this.type = type;
        this.value = new IntegerTemplate(value);
        this.ignoreShield = ignoreShield;
    }

    public DamageType getType()
    {
        return type;
    }

    public IntegerTemplate getValueTemplate()
    {
        return value;
    }

    public int getValue()
    {
        return value.getValue();
    }

    public boolean ignoreShield()
    {
        return ignoreShield;
    }

    public static enum DamageType
    {
        PHYSIC,
        MAGIC,
        TRUE;
    }
}

File: src/main/java/io/myosotisdev/utopianism/battle/DamageModifier.kt
package io.myosotisdev.utopianism.battle

import io.myosotisdev.utopianism.util.IntegerTemplate

class DamageModifier @JvmOverloads constructor(val source: String, val type: DamageType, value: Int, ignoreShield: Boolean = false)
{
    val template: IntegerTemplate
    val isIgnoreShield: Boolean

    init
    {
        template = IntegerTemplate(value)
        isIgnoreShield = ignoreShield
    }

    fun getValue(): Int
    {
        return template.value()
    }

    fun ignoreShield(): Boolean
    {
        return isIgnoreShield
    }

    enum class DamageType
    {
        PHYSIC,
        MAGIC,
        TRUE
    }
}